Lib: Debugs for route-map code in FRR

adharkar opened the pull request:
Added a CLI "debug route-map" to enble route-map debugs

Added debugs for following triggers
1. Add/delete a route-map
2. Add/delete a sequence in route-map
3. Add/delete a match statement(dependency)
4. Update a dependency
5. Apply a route-map

Signed-off-by: Ameya Dharkar <adharkar@vmware.com>

donaldsharp commented on bgpd/bgp_route.c:
```diff
@@ -1781,6 +1781,10 @@ int subgroup_announce_check(struct bgp_node *rn, struct bgp_path_info *pi,
 		peer->rmap_type = 0;
 
 		if (ret == RMAP_DENYMATCH) {
+			if (bgp_debug_update(NULL, p, subgrp->update_group, 0))
```
The changes for the lib/routemap.c code are sufficient enough that this needs to be in a different commit.

adharkar replied:
Okay..I will open another pull request for this.

donaldsharp commented on lib/routemap.c:
```diff
@@ -677,7 +677,10 @@ static unsigned int route_map_dep_hash_make_key(void *p);
-static int rmap_debug = 0;
+bool rmap_debug;
+
+#define RMAP_DEBUG true
```
just remove this and #define IS_RMAP_DEBUG (rmap_debug)

donaldsharp replied:
Additionally if we are going to have a IS_RMAP_DEBUG define we either need to use it consistently or not.  Pick one for your tests.

donaldsharp commented on lib/routemap.c:
```diff
@@ -1542,18 +1560,23 @@ route_map_result_t route_map_apply(struct route_map *map,
 			RMAP_RECURSION_LIMIT);
 		recursion = 0;
-		return RMAP_DENYMATCH;
+		ret = RMAP_DENYMATCH;
```
Why do we need to have another debug here?  Isn't the flog_warn enough?  I say just leave the return RMAP_DENYMATCH

adharkar replied:
Fixed

donaldsharp commented on lib/routemap.c:
```diff
@@ -2867,6 +2909,31 @@ DEFUN (no_rmap_description,
+       DEBUG_STR
+       "Debug option set for route-maps\n")
+{
+	rmap_debug = RMAP_DEBUG;
```
true

adharkar replied:
Fixed

donaldsharp commented on lib/routemap.c:
```diff
@@ -2912,6 +2979,18 @@ static int route_map_config_write(struct vty *vty)
+		write++;
+	}
+
+	return (write);
```
return write;

louberger commented:
```

  CCLD     pbrd/pbrd
  CCLD     staticd/staticd
  CCLD     bfdd/bfdd
  AR       bgpd/libbgp.a
  CCLD     ospfclient/ospfclient
lib/.libs/libfrr.a(routemap.o):(.data+0x420): multiple definition of `debug_node'
zebra/debug.o:(.data+0x20): first defined here
collect2: ld returned 1 exit status
make[1]: *** [zebra/zebra] Error 1
make[1]: *** Waiting for unfinished jobs....

```

sarav511 commented on lib/routemap.c:
```diff
@@ -677,7 +677,7 @@ static unsigned int route_map_dep_hash_make_key(void *p);
 static void route_map_rule_delete(struct route_map_rule_list *,
 				  struct route_map_rule *);
-static int rmap_debug = 0;
+bool rmap_debug;
```
Shall I know the intention behind removing static keyword?

donaldsharp commented on lib/routemap.c:
```diff
@@ -1053,6 +1059,10 @@ static void route_map_index_delete(struct route_map_index *index, int notify)
 
+	if (rmap_debug)
+		zlog_debug("Deleting route-map %s sequence %d",
+				index->map->name, index->pref);
```
tab/spaces look wrong here

donaldsharp commented on lib/routemap.c:
```diff
@@ -1141,6 +1151,11 @@ route_map_index_add(struct route_map *map, enum route_map_type type, int pref)
+
+	if (rmap_debug)
+		zlog_debug("Route-map %s add sequence %d, type: %s",
+			map->name, pref, RMAP_TYPE_STR(type));
```
again tab/spaces are wrong.

donaldsharp commented on lib/routemap.c:
```diff
@@ -1562,6 +1581,13 @@ route_map_result_t route_map_apply(struct route_map *map,
 
+		if (rmap_debug) {
+			zlog_debug("Route-map: %s, sequence: %d, prefix: %s, result: %s",
+				map->name, index->pref,
```
tab spaces are wrong

donaldsharp commented on lib/routemap.h:
```diff
@@ -37,6 +37,9 @@ DECLARE_MTYPE(ROUTE_MAP_COMPILED)
 /* Route map's type. */
 enum route_map_type { RMAP_PERMIT, RMAP_DENY, RMAP_ANY };
 
+#define RMAP_TYPE_STR(type) ((type == RMAP_PERMIT) ? "permit" : \
```
Isn't this route_map_type_str()?  Why do we need another one?

donaldsharp commented on lib/routemap.h:
```diff
@@ -45,6 +48,12 @@ typedef enum {
 	RMAP_OKAY
 } route_map_result_t;
 
+#define RMAP_RESULT_STR(res) ((res == RMAP_MATCH) ? "match" : \
```
So I would prefer we follow the standard we have throughout the code base for enum to strings and have a specialized function that has a switch statement ( see route_map_type_str() ) in that if we add a new value we will have a failed compile instead of having to `know` to edit this particular #define.  Especially since we don't follow this pattern any where else in the code base

donaldsharp commented on lib/routemap.c:
```diff
@@ -1616,19 +1642,30 @@ route_map_result_t route_map_apply(struct route_map *map,
+route_map_apply_end:
+	if (rmap_debug) {
+		zlog_debug("Route-map: %s, prefix: %s, result: %s",
+				(map ? map->name : "null"),
```
Fix tab/spaces

adharkar commented:
> ## 🚧 Basic BGPD CI results: Partial FAILURE, 110 tests failed, has VALGRIND issues
> Results table
> _	_
> Result	SUCCESS git merge/4300 [253c6ec](https://github.com/FRRouting/frr/commit/253c6eced0fcb6387a2c312645259d92b8d39d3a)
> Date	05/15/2019
> Start	16:30:22
> Finish	16:54:47
> Run-Time	24:25
> Total	1915
> Pass	1805
> Fail	110
> Valgrind-Errors	0
> Valgrind-Loss	17
> Details	vncregress-2019-05-15-16:30:22.txt
> Log	autoscript-2019-05-15-16:31:15.log.bz2
> Memory	424 429 359
> For details, please contact louberger

Hi, can I get the details of these failures?

srimohans commented:
ci:rerun

srimohans commented:
CI:rerun

donaldsharp commented:
@louberger -> what is the cause of your CI failure here?  I do not see how what was done can have valgrind issues.

qlyoung commented on lib/routemap.c:
```diff
@@ -1616,19 +1660,30 @@ route_map_result_t route_map_apply(struct route_map *map,
+	if (rmap_debug) {
+		zlog_debug("Route-map: %s, prefix: %s, result: %s",
+			   (map ? map->name : "null"),
+			   prefix2str(prefix, buf, sizeof(buf)),
```
if I'm not mistaken we don't need `buf` and `buf2`, we can just use one `buf` right?

qlyoung commented on lib/routemap.c:
```diff
@@ -2867,6 +2923,30 @@ DEFUN (no_rmap_description,
+
+DEFUN (no_debug_rmap,
+       no_debug_rmap_cmd,
+       "no debug route-map",
```
Could you please collapse this DEFUN into the previous one?
```
DEFUN (debug_rmap,
       debug_rmap_cmd,
       "[no] debug route-map",
       NO_STR
       DEBUG_STR
       "Debug option set for route-maps\n")
{
	rmap_debug = !no;
	return CMD_SUCCESS;
}
```

adharkar replied:
This gives following error.

lib/routemap.c: In function 'debug_rmap':
lib/routemap.c:3071:16: error: 'no' undeclared (first use in this function)
  rmap_debug = !no;
                ^
lib/routemap.c:3071:16: note: each undeclared identifier is reported only once f

Is there any other way of doing this?

for all other debug commands, [no] is handled in a  separate command.

qlyoung replied:
sorry, should have wrote `DEFPY` not `DEFUN`

louberger commented:
here are the valgrind errors:
```
==1618== HEAP SUMMARY:
==1618==     in use at exit: 509 bytes in 17 blocks
==1618==   total heap usage: 278,141 allocs, 278,124 frees, 13,629,436 bytes allocated
==1618== 
==1618== Searching for pointers to 17 not-freed blocks
==1618== Checked 559,904 bytes
==1618== 
==1618== 8 bytes in 1 blocks are still reachable in loss record 1 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x4E3EAF: graph_new (graph.c:32)
==1618==    by 0x4DBFD8: install_node (command.c:353)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:graph_new
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 8 bytes in 1 blocks are still reachable in loss record 2 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x503E18: vector_init (vector.c:40)
==1618==    by 0x4E3EBC: graph_new (graph.c:33)
==1618==    by 0x4DBFD8: install_node (command.c:353)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:vector_init
   fun:graph_new
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 8 bytes in 1 blocks are still reachable in loss record 3 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x503E18: vector_init (vector.c:40)
==1618==    by 0x4DBFE6: install_node (command.c:354)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:vector_init
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 8 bytes in 1 blocks are still reachable in loss record 4 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x503E18: vector_init (vector.c:40)
==1618==    by 0x4E3EF6: graph_new_node (graph.c:53)
==1618==    by 0x4DC00B: install_node (command.c:358)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:vector_init
   fun:graph_new_node
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 8 bytes in 1 blocks are still reachable in loss record 5 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x503E18: vector_init (vector.c:40)
==1618==    by 0x4E3F03: graph_new_node (graph.c:54)
==1618==    by 0x4DC00B: install_node (command.c:358)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:vector_init
   fun:graph_new_node
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 13 bytes in 1 blocks are still reachable in loss record 6 of 17
==1618==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
==1618==    by 0x5F1A0C9: strdup (in /usr/lib64/libc-2.17.so)
==1618==    by 0x4EAE13: qstrdup (memory.c:122)
==1618==    by 0x4E463F: hash_create_size (hash.c:54)
==1618==    by 0x4DC024: install_node (command.c:360)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:malloc
   fun:strdup
   fun:qstrdup
   fun:hash_create_size
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 16 bytes in 1 blocks are still reachable in loss record 7 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x503DEB: vector_init (vector.c:32)
==1618==    by 0x4E3EBC: graph_new (graph.c:33)
==1618==    by 0x4DBFD8: install_node (command.c:353)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:vector_init
   fun:graph_new
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 16 bytes in 1 blocks are still reachable in loss record 8 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x503DEB: vector_init (vector.c:32)
==1618==    by 0x4DBFE6: install_node (command.c:354)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:vector_init
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 16 bytes in 1 blocks are still reachable in loss record 9 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x503DEB: vector_init (vector.c:32)
==1618==    by 0x4E3EF6: graph_new_node (graph.c:53)
==1618==    by 0x4DC00B: install_node (command.c:358)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:vector_init
   fun:graph_new_node
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 16 bytes in 1 blocks are still reachable in loss record 10 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x503DEB: vector_init (vector.c:32)
==1618==    by 0x4E3F03: graph_new_node (graph.c:54)
==1618==    by 0x4DC00B: install_node (command.c:358)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:vector_init
   fun:graph_new_node
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 24 bytes in 1 blocks are still reachable in loss record 11 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x4E88A2: listnode_add (linklist.c:59)
==1618==    by 0x4E467D: hash_create_size (hash.c:62)
==1618==    by 0x4DC024: install_node (command.c:360)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:listnode_add
   fun:hash_create_size
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 32 bytes in 1 blocks are still reachable in loss record 13 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x4E3EE9: graph_new_node (graph.c:50)
==1618==    by 0x4DC00B: install_node (command.c:358)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:graph_new_node
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 40 bytes in 1 blocks are still reachable in loss record 14 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x4E4667: hash_create_size (hash.c:60)
==1618==    by 0x5023D3: thread_master_create (thread.c:435)
==1618==    by 0x4E823C: frr_init (libfrr.c:636)
==1618==    by 0x429EFC: main (bgp_main.c:456)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:hash_create_size
   fun:thread_master_create
   fun:frr_init
   fun:main
}
==1618== 64 bytes in 1 blocks are still reachable in loss record 15 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x4E4602: hash_create_size (hash.c:47)
==1618==    by 0x4DC024: install_node (command.c:360)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:hash_create_size
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 72 bytes in 1 blocks are still reachable in loss record 16 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x4DE362: cmd_token_new (command_graph.c:38)
==1618==    by 0x4DBFFA: install_node (command.c:356)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:cmd_token_new
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== 128 bytes in 1 blocks are still reachable in loss record 17 of 17
==1618==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==1618==    by 0x4EAD92: qcalloc (memory.c:110)
==1618==    by 0x4E4616: hash_create_size (hash.c:49)
==1618==    by 0x4DC024: install_node (command.c:360)
==1618==    by 0x437022: bgp_debug_init (bgp_debug.c:2239)
==1618==    by 0x4935E1: bgp_init (bgpd.c:8018)
==1618==    by 0x429F74: main (bgp_main.c:470)
==1618== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:calloc
   fun:qcalloc
   fun:hash_create_size
   fun:install_node
   fun:bgp_debug_init
   fun:bgp_init
   fun:main
}
==1618== LEAK SUMMARY:
==1618==    definitely lost: 0 bytes in 0 blocks
==1618==    indirectly lost: 0 bytes in 0 blocks
==1618==      possibly lost: 0 bytes in 0 blocks
==1618==    still reachable: 477 bytes in 16 blocks
==1618==         suppressed: 32 bytes in 1 blocks
==1618== 
```

adharkar commented:
CI:rerun

adharkar commented:
Hi @donaldsharp @qlyoung , I have fixed the memory leak issue. I t was because of incorrect definition of the debug_node.
I have also addressed review comments. Can this be merged now?

Regards!
